[mlir][llvm] Support nusw and nuw in GEP

lpy opened the pull request:
nusw and nuw were introduced in getelementptr, this patch plumbs them in MLIR.

Since inbounds implies nusw, this patch also adds an inboundsFlag to represent the concept of raw inbounds with no nusw implication, and have the inbounds literal captured as the combination of inboundsFlag and nusw.

Fixes: iree#20482

lpy committed:
Update MLIR to support nusw and nuw in GEP.

nusw and nuw were introduced in getelementptr, this patch plumbs them in MLIR.

Since inbounds implies nusw, this patch also adds an inboundsFlag to
represent the concept of raw inbounds with no nusw implication, and have
the inbounds literal captured as the combination of inboundsFlag and
nusw.

Fixes: iree-org/iree#20482
Signed-off-by: Lin, Peiyong <linpyong@gmail.com>

gysit commented on mlir/lib/Target/LLVMIR/ModuleImport.cpp:
```diff
@@ -1984,8 +1984,9 @@ LogicalResult ModuleImport::convertInstruction(llvm::Instruction *inst) {
-                                       indices, gepInst->isInBounds());
+    auto gepOp = builder.create<GEPOp>(
+        loc, type, sourceElementType, *basePtr, indices,
+        gepInst->isInBounds() ? GEPNoWrapFlags::inbounds : GEPNoWrapFlags());
```
Can you import all the no wrap flags here and add a test to ensure the import works?

A test could be added here:
`llvm-project/mlir/test/Target/LLVMIR/Import/instructions.ll` or possibly even in a separate file.

gysit commented on mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:
```diff
@@ -291,7 +291,7 @@ def LLVM_GEPOp : LLVM_Op<"getelementptr", [Pure,
                    DenseI32ArrayAttr:$rawConstantIndices,
                    TypeAttr:$elem_type,
-                   UnitAttr:$inbounds);
+                   GEPNoWrapFlagsProp:$gepNoWrapFlags);
```
```suggestion
                   GEPNoWrapFlagsProp:$noWrapFlags);
```
nit: I would probably skip the gep prefix.

gysit commented on mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:
```diff
@@ -303,8 +303,10 @@ def LLVM_GEPOp : LLVM_Op<"getelementptr", [Pure,
     either use constant indices directly, or supply a constant SSA value.
 
-    An optional 'inbounds' attribute specifies the low-level pointer arithmetic
+    Optional attributes can be used to specify the low-level pointer arithmetic
```
```suggestion
    The no-wrap flags can be used to specify the low-level pointer arithmetic
```
nit: I would name the flag explicitly.

gysit commented on mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:
```diff
@@ -303,8 +303,10 @@ def LLVM_GEPOp : LLVM_Op<"getelementptr", [Pure,
-    An optional 'inbounds' attribute specifies the low-level pointer arithmetic
+    Optional attributes can be used to specify the low-level pointer arithmetic
     overflow behavior that LLVM uses after lowering the operation to LLVM IR.
+    The acceptable attributes could be one of or the combination of 'inbounds',
+    'nusw' or 'nuw'.
```
Could it make sense to provide a very short definition of the flags (I may be off in terms of the actual semantics):

Valid options include 'inbounds' (pointer arithmetic must be within object bounds),
'nsw' (no signed wrap), and 'nuw' (no unsigned wrap). The flags can be set individually or in combination.

krzysz00 replied:
I'll not it's `nusw` - no unsigned signed wrap . And the separate note that `inbounds` implies `nusw`, which the enum definition ensures

gysit replied:
@krzysz00 thanks for the correction! I did edit the text above. Hopefully it is correct now.

joker-eph replied:
How is it ensured by the enum definition? I would expect that this means that the enum is a bitfield and the `inbounds` has the bit set for both positions?

joker-eph replied:
Ah I misread: this is actually done (I had only seen `GEPInboundsFlag` instead of `GEPInbounds` which does the grouping).

joker-eph replied:
What happens if someone uses GEPInboundsFlag to build the operation?

krzysz00 replied:
I think we need a verifier somewhere (probably on the property or on the op) to catch inbounds_flag without nusw 

Currently, I think that combination goes to an LLVM assertion in the translator

lpy replied:
If `inbounds_flag` is used the assertion will be triggered in LLVM: https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/IR/GEPNoWrapFlags.h#L35

gysit replied:
Yeah I would also be in favor of a verifier in MLIR itself. Triggering an assert after exporting to LLVM IR is not ideal.

gysit reviewed (commented):
Thanks for fixing this!

Can you make sure we import the flags properly into MLIR?

lpy commented:
Thanks for the comments. I updated the patch, please take another look :)

Dinistro reviewed (approved):
Maybe also add this to the LLVM dialect's roundtrip test, just to be sure that the parsing and printing works as expected. Apart from that, this is LGTM, thanks for the patch.

gysit reviewed (commented):
Thanks for addressing the comments.

Generally LGTM, but I think would make sense to add a verifier that ensures the inbounds flag never shows up on its own.

lpy commented:
Thanks for the feedback. I have added a few lines to the roundtrip test and updated the `verify` function to capture the usage of `inbounds_flag` in early stage. Please take a look :)

joker-eph commented on mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp:
```diff
@@ -794,6 +794,9 @@ LogicalResult LLVM::GEPOp::verify() {
            << getRawConstantIndicesAttrName().getValue() << "'";
 
+  if (getNoWrapFlags() == GEPNoWrapFlags::inboundsFlag)
+    return emitOpError("'inbounds_flag' can not be used directly.");
```
```suggestion
    return emitOpError("'inbounds_flag' cannot be used directly.");
```

We also need a test for this in the `invalid.mlir` test file.

joker-eph commented on mlir/test/Dialect/LLVMIR/roundtrip.mlir:
```diff
@@ -236,6 +236,14 @@ llvm.func @gep(%ptr: !llvm.ptr, %idx: i64, %ptr2: !llvm.ptr) {
   // CHECK: llvm.getelementptr inbounds %{{.*}}[%{{.*}}, 0, %{{.*}}] : (!llvm.ptr, i64, i64) -> !llvm.ptr, !llvm.struct<(array<10 x f32>)>
   llvm.getelementptr inbounds %ptr2[%idx, 0, %idx] : (!llvm.ptr, i64, i64) -> !llvm.ptr, !llvm.struct<(array<10 x f32>)>
+  // CHECK: llvm.getelementptr inbounds|nuw %{{.*}}[%{{.*}}, 0, %{{.*}}] : (!llvm.ptr, i64, i64) -> !llvm.ptr, !llvm.struct<(array<10 x f32>)>
+  llvm.getelementptr inbounds | nuw %ptr2[%idx, 0, %idx] : (!llvm.ptr, i64, i64) -> !llvm.ptr, !llvm.struct<(array<10 x f32>)>
```
can you add one "inbounds nsw" case? It should parse and print "inbounds" only I believe.

lpy commented:
Thanks, added a case to cover that. Please take a look :)

gysit reviewed (approved):
LGTM

Let's give the other reviewers a chance to look at it before landing.

krzysz00 reviewed (approved):
I think this has all the tests that were called for, lgtm, we're probably good to merge tomorrow?
